subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java

```diff
@@ -45,4 +62,71 @@ public boolean execute(Node node, NodeExecutionContext context) {
             return false;
         }
     }
+
+    private void detectMissingDependencies(LocalTaskNode node) {
+        RelatedLocations consumedDirectories = consumedAndProducedLocations.getConsumedDirectories();
+        for (String outputPath : node.getMutationInfo().outputPaths) {
+            consumedDirectories.getNodesRelatedTo(outputPath).stream()
+                .filter(consumerNode -> missesDependency(node, consumerNode))
+                .forEach(consumerWithoutDependency -> emitMissingDependencyDeprecationWarning(node, consumerWithoutDependency));
+        }
+        Set<String> consumedLocations = new LinkedHashSet<>();
```

Comment:
So we report the errors in the same order. I suppose we can use a regular `HashSet`, now that we use validation context, since the messages will be sorted there?